js/fetch.ts

```diff
@@ -97,59 +99,127 @@ export const Headers = DomIterableMixin<string, string, typeof HeadersBase>(
+        this._chunks = [];
+        this._bodyWaiter.resolve(this._data!.buffer as ArrayBuffer);
+      } catch (e) {
+        this._bodyWaiter.reject(e);
+      } finally {
+        this._buffering = false;
+      }
+    }
+  }
+
+  async arrayBuffer(): Promise<ArrayBuffer> {
+    this._bufferBody();
+    return this._bodyWaiter;
+  }
+
+  async blob(): Promise<domTypes.Blob> {
+    const arrayBuffer = await this.arrayBuffer();
+    return new DenoBlob([arrayBuffer], {
+      type: this.contentType
+    });
+  }
+
+  async formData(): Promise<domTypes.FormData> {
+    notImplemented();
+    return {} as domTypes.FormData;
+  }
+
+  // tslint:disable-next-line:no-any
+  async json(): Promise<any> {
+    const text = await this.text();
+    return JSON.parse(text);
+  }
+
+  async text(): Promise<string> {
+    const ab = await this.arrayBuffer();
+    const decoder = new TextDecoder("utf-8");
+    return decoder.decode(ab);
+  }
+
+  read(p: ArrayBufferView): Promise<io.ReadResult> {
+    return read(this.rid, p);
+  }
+
+  async cancel(): Promise<void> {
+    notImplemented();
+  }
+
+  getReader(): domTypes.ReadableStreamReader {
+    notImplemented();
+    return {} as domTypes.ReadableStreamReader;
+  }
+}
+
+class Response implements domTypes.Response {
   readonly url: string = "";
-  body: null;
+  body: Body;
   bodyUsed = false; // TODO
   statusText = "FIXME"; // TODO
   readonly type = "basic"; // TODO
   redirected = false; // TODO
   headers: domTypes.Headers;
   readonly trailer: Promise<domTypes.Headers>;
-  //private bodyChunks: Uint8Array[] = [];
-  private first = true;
-  private bodyData: ArrayBuffer;
-  private bodyWaiter: Resolvable<ArrayBuffer>;
 
   constructor(
     readonly status: number,
-    readonly body_: ArrayBuffer,
-    headersList: Array<[string, string]>
+    headersList: Array<[string, string]>,
+    private bodyRid: number
   ) {
-    this.bodyWaiter = createResolvable();
     this.trailer = createResolvable();
     this.headers = new Headers(headersList);
-    this.bodyData = body_;
-    setTimeout(() => {
-      this.bodyWaiter.resolve(body_);
-    }, 0);
+    const contentType = this.headers.get("content-type") || "";
+    this.body = new Body(bodyRid, contentType);
   }
 
-  arrayBuffer(): Promise<ArrayBuffer> {
-    return this.bodyWaiter;
+  async arrayBuffer(): Promise<ArrayBuffer> {
+    return this.body.arrayBuffer();
   }
 
   async blob(): Promise<domTypes.Blob> {
-    const arrayBuffer = await this.arrayBuffer();
-    return new DenoBlob([arrayBuffer], {
-      type: this.headers.get("content-type") || ""
-    });
+    return this.body.blob();
   }
 
   async formData(): Promise<domTypes.FormData> {
-    notImplemented();
-    return {} as domTypes.FormData;
+    return this.body.formData();
   }
 
-  async json(): Promise<object> {
-    const text = await this.text();
-    return JSON.parse(text);
+  // tslint:disable-next-line:no-any
+  async json(): Promise<any> {
```

Comment:
the type to use here would be `unknown`.

Reply:
Same error: https://github.com/denoland/deno/pull/1102#discussion_r228714933